### DISCUSSION TITLE
FEM on manifold or shell

### DISCUSSION THREAD

**Discussion by gaelgrissonnanche** [kinnala/scikit-fem#1039] [General]
Hi,

I wonder if it is possible to do FEM on a shell (or manifold), I mean FEM on a surface in a 3D space with scikit-fem?
I know it is possible with FEniCS-shell https://fenics-shells.readthedocs.io/en/latest/

Thanks!
Gaël

**Comment by kinnala** [comment]
Can you give an example problem? I have solved some of the problems given in FEniCS-shell documentation including shell problems where a parametrization of the midsurface is available. I have not solved problems where the surface is given by a triangle mesh embedded in R^3 but I think it should be possible.

**Reply by kinnala** [reply]
Here I tried discretizing the Laplace-Beltrami operator on the surface of a sphere using the ideas laid out in the following paper: https://link.springer.com/article/10.1007/s00211-018-0990-2

```python
from skfem import *
from skfem.helpers import *

m = MeshTet.init_ball(nrefs=3)

basis = FacetBasis(m, ElementTetP1())


@BilinearForm
def laplace(u, v, w):
    I = identity(w.n)
    nxn = prod(w.n, w.n)
    return dot(mul(I - nxn, grad(u)), mul(I - nxn, grad(v)))


@LinearForm
def periodic(v, w):
    theta = np.arctan2(w.x[1], w.x[0])
    return np.sin(2. * np.pi * theta) * v


A = laplace.assemble(basis)
b = periodic.assemble(basis)

y = solve(*condense(A, b, I=basis.get_dofs()))


m.save('lapbel.vtk', point_data={'y': y})
```

I didn't do any convergence checks but qualitatively it seems to give a reasonable result. It is using the surface mesh representation.

<img width="513" alt="Screenshot 2023-07-10 at 12 36 53" src="https://github.com/kinnala/scikit-fem/assets/973268/14c2aca7-8fb9-4ebb-9649-408d352a87e9">

**Reply by kinnala** [reply]
Note that there isn't currently a mesh structure for embedded triangle meshes in scikit-fem. As a workaround there has to be a volume mesh whose boundary is the triangle mesh that is of interest. In the above example I generated a volume mesh for a ball and used its surface mesh for the finite element computations on the surface of a sphere. If you have just the triangle mesh you could extend each triangle element by one node to turn them into tetrahedra.
